### ISSUE TITLE
[Modded] Exception: Object reference not set to an instance of an object in the Vintagestory.GameContent.Mechanics.MechanicalPowerMod.RebuildNetwork

### ISSUE THREAD

**Issue by fipil** [anegostudios/VintageStory-Issues#3663]
### Game Version

v1.19.4

### Platform

Windows

### Modded

Modded, mods installed: creaturescan, extrachests, HangingOilLamps, primitivesurvival, rope-bridges, game, weaponpackalphaunoff, carryon, commonlib, djinlandessentials, electricity, sailboat, medievalexpansion, morebags, mcrate, petai, rivers, stonerailings, trailmod, tstools, vinconomy, creative, survival, workbenchexpansion, ageofconfession, feverstonehorses, stonequarry, wolftaming

### SP/MP

Multiplayer

### Description

Sometimes when I connect to the MP server, I notice that some chunk column(s) around me is not loaded, there is just a hole and I cannot pass in, the chunk column is never loaded. Lokking into the server console I see there the following error:
```
3.3.2024 11:51:05 [Server Event] fipil 84.42.170.155:62172 joins.
3.3.2024 11:51:05 [Server Error] Exception: Object reference not set to an instance of an object.
   at Vintagestory.GameContent.Mechanics.MechanicalPowerMod.RebuildNetwork(MechanicalNetwork network, IMechanicalPowerDevice n
owRemovedNode) in VSSurvivalMod\Systems\MechanicalPower\MechanicalPowerMod.cs:line 233
   at Vintagestory.GameContent.Mechanics.MechanicalPowerMod.Event_ChunkDirty(Vec3i chunkCoord, IWorldChunk chunk, EnumChunkDir
tyReason reason) in VSSurvivalMod\Systems\MechanicalPower\MechanicalPowerMod.cs:line 289
   at Vintagestory.Server.ServerSystemSupplyChunks.mainThreadLoadChunkColumn(ChunkColumnLoadRequest chunkRequest) in Vintagest
oryLib\Server\Systems\World\LoadThread\SupplyChunks.cs:line 890
   at Vintagestory.Server.ServerSystemSupplyChunks.<>c__DisplayClass15_0.<loadOrGenerateChunkColumn_OnChunkThread>b__0() in Vi
ntagestoryLib\Server\Systems\World\LoadThread\SupplyChunks.cs:line 692
   at Vintagestory.Server.ServerMain.ProcessMainThreadTasks() in VintagestoryLib\Server\ServerMain.cs:line 2782
   at Vintagestory.Server.ServerMain.ProcessMain() in VintagestoryLib\Server\ServerMain.cs:line 893
   at Vintagestory.Server.ServerMain.Process() in VintagestoryLib\Server\ServerMain.cs:line 886
3.3.2024 11:51:05 [Server Notification] A client reconnected, resuming game calendar.
```
I have to restart the server and login again to solve the problem. It doesn't happen regularly, just randomly.
It started to happen after we upgraded server from 1.19.4. We used the db prune after the upgrade and we run the db prune on the 1.19.3 version because db prune in the 1.19.4 version was crashing. Maybe this is related to the problem, maybe not?

### Logs

```Log```

**Comment by Craluminum2413** [comment]
What mods are installed?

**Comment by fipil** [comment]
We use those mods:
mod@creaturescan_1.0.6.zip, mod@ExtraChests_v1.8.0.zip, mod@hangingoillamps_1.0.1.zip, mod@primitivesurvival_3.4.7.zip, mod@RopeBridges1.1.0forVS1.15.zip, mod@WeaponPackAlphaUnofficial_1.5.2.zip, mod@CarryOn-1.19_v1.7.4.zip, mod@DjinlandEssentials.1.2.2.zip, mod@electricity_0.0.11.zip, mod@josailing-1.2.8.zip, mod@medievalexpansion-3.13.1.zip, mod@MoreBags-1.2.0.zip, mod@MultiblockCrates-v1.2.1.zip, mod@petai_v2.2.3.zip, mod@rivers_2.7.0.zip, mod@StoneRailings.1.2.1.zip, mod@trailmod_1.0.7.zip, mod@vinconomy_0.2.3.zip, mod@workbench-expansion-1.8.0.zip, mod@AgeOfConfession.zip, mod@FeverstoneHorses-v1.6.0-1.19.0.zip, mod@StoneQuarry_VS1.19.0-rc.5_net7_v3.2.1.zip, mod@wolftaming_v2.1.0_petai_v2.0.+.zip, mod@tstools_v2.0.4.zip

**Comment by Craluminum2413** [comment]
@fipil Does it happen without any mods?

**Comment by fipil** [comment]
As it happens randomly. I don't know. I'll try it locally on a map copy an I'll let you know.

**Comment by fipil** [comment]
@Craluminum2413 it happend once when testing locally, with all mods. Then I tried it without mods and it did not happen. 
So maybe it's some mod. I'll try to remove some mods and watch if the error happends again or not on server.

**Comment by radfast** [comment]
In the vanilla game, relevant code has not been updated at all since game version 1.18.  Mechanical Power is essentially the same in 1.19.x, so I don't think it's something new in 1.19.4 specifically.  Could be the server changed mods at the same time as updating to 1.19.4?

The line numbers in the crash / stack trace do not match the line numbers in the game's source code, at all (one of them is 21 lines out), which does strongly suggest that one of the mods has made code changes to Mechanical Power here.  I don't know _which_ mod, but must be one of the **Code** mods.

**Comment by fipil** [comment]
> 
> The line numbers in the crash / stack trace do not match the line numbers in the game's source code, at all (one of them is 21 lines out), which does strongly suggest that one of the mods has made code changes to Mechanical Power here. I don't know _which_ mod, but must be one of the **Code** mods.

Yes, I noticed that line mismatch, it's confusing. I was investigating if any of mods used by us does use the Harmony for patching of those methods, but didn't find any. Now I suspect the mod Electricity (https://mods.vintagestory.at/electricity), which is adding the Motors implementing IMechanicalPowerBlock to game. I was using a couple of the Motor blocks in the problematic chunk. But since the bug is so rare it's hard to catch. I tried to catch it in debug run, but no success. 
Now I moved all those Motors to a different location and built a slightly different machine configurations and the exception is gone. 

So, maybe we can close the issue for now.  Thanks.

**Comment by DragonDev1906** [comment]
We have one chunk that goes missing regularly, too, with the same exception in the logs. This one happened while someone was joining inside the chunk that has gone missing (checked with coordinates in debug menu).

```
31.3.2024 22:57:00 [Notification] Client uid **** attempting identification. Name: ****
31.3.2024 22:57:02 [Notification] Delayed join, need to load one spawn chunk first.
31.3.2024 22:57:02 [Notification] Sending server identification with remap False.  Server control privilege is True
31.3.2024 22:57:02 [Error] Exception: Object reference not set to an instance of an object.
   at Vintagestory.GameContent.Mechanics.MechanicalPowerMod.RebuildNetwork(MechanicalNetwork network, IMechanicalPowerDevice nowRemovedNode) in VSSurvivalMod\Systems\MechanicalPower\MechanicalPowerMod.cs:line 233
   at Vintagestory.GameContent.Mechanics.MechanicalPowerMod.Event_ChunkDirty(Vec3i chunkCoord, IWorldChunk chunk, EnumChunkDirtyReason reason) in VSSurvivalMod\Systems\MechanicalPower\MechanicalPowerMod.cs:line 289
   at Vintagestory.Server.ServerSystemSupplyChunks.mainThreadLoadChunkColumn(ChunkColumnLoadRequest chunkRequest) in VintagestoryLib\Server\Systems\World\LoadThread\SupplyChunks.cs:line 894
   at Vintagestory.Server.ServerSystemSupplyChunks.<>c__DisplayClass15_0.<loadOrGenerateChunkColumn_OnChunkThread>b__0() in VintagestoryLib\Server\Systems\World\LoadThread\SupplyChunks.cs:line 696
   at Vintagestory.Server.ServerMain.ProcessMainThreadTasks() in VintagestoryLib\Server\ServerMain.cs:line 2782
   at Vintagestory.Server.ServerMain.ProcessMain() in VintagestoryLib\Server\ServerMain.cs:line 893
   at Vintagestory.Server.ServerMain.Process() in VintagestoryLib\Server\ServerMain.cs:line 886
31.3.2024 22:57:02 [Notification] Placing player at 509875.6899299805 129 513891.794425293
31.3.2024 22:57:40 [Event] **** joins.
```

Interestingly this lead to (almost?) no chunks being visible for him except for the chunk he is in, which we didn't see:
![image](https://github.com/anegostudios/VintageStory-Issues/assets/8270201/7fae11b6-0d2a-4fb7-aaca-9bf446dbedb3)

Mods loaded: mod@ACulinaryArtillery 1.1.3.zip, mod@AncientArmory0.9.5.zip, mod@BetterCrates_v1.7.0.zip, mod@BetterTradersv0.0.5.zip, mod@chiseltools1.10.8.zip, mod@decorbazaar_v1.0.15.zip, mod@ExtraChests_v1.8.0.zip, mod@FotSA-Caninae-v1.0.10.zip, mod@FotSA-Capreolinae-v1.0.5.zip, mod@FotSA-Machairodontinae-v1.0.3.zip, mod@FotSA-Pantherinae-v1.1.1.zip, mod@FotSA-Sirenia-v1.0.0.zip, mod@GeologyAdditions 1.3.3.zip, mod@Mannequin-Stand-v1.0.3.zip, mod@millwright_1.1.4.zip, mod@Plainsandvalleys-v1.0.5.zip, mod@primitivesurvival_3.5.0.zip, mod@StonePathVariants-v1.0.5.zip, mod@viescraftmachines_2.3.1.zip, mod@ABCSReborn-Hardcore-v.0.1.7.zip, mod@AncientTools_V1.5.17.zip, mod@beehivekiln_1.5.1.zip, mod@betterjonasdevices_1.1.0.zip, mod@BetterRuinsv0.3.4.zip, mod@CarryOn-1.19_v1.7.4.zip, mod@electricity_0.0.11.zip, mod@ExpandedFoods 1.6.8.zip, mod@joyofsailing-1.3.0.zip, mod@medievalexpansion-3.13.1.zip, mod@nbcartographer_2.0.9.zip, mod@PlayerList-1.8.0.zip, mod@ProspectTogether-1.3.0.zip, mod@rivers_2.5.0.zip, mod@StoneRailings.1.2.1.zip, mod@StorageOptions-v1.0.2.zip, mod@StructuresSpawnChances200%.zip, mod@survivalnutrition_1.0.0.zip, mod@thecritterspackage090.zip, mod@undergroundmines_1.0.3.zip, mod@AWearableLight-v1.1.1.zip, mod@cob_tradercamps_1.1.3.zip, mod@expanded_matter-2.5.2.zip, mod@HelveHammerExtensions_VS1.19.0-rc.5_net7_v1.6.0.zip, mod@PlayerCorpse_VS1.19.0-rc.5_net7_v1.9.0.zip, mod@wildcraftfruit_1.1.1.zip, mod@wildcraftherb_0.0.0.zip, mod@wildcrafttree_1.1.0.zip, mod@bricklayers-2.5.1.zip

The following mods are on both installs:
```
# Shared
mod@ExtraChests_v1.8.0.zip
mod@primitivesurvival_3.5.0.zip				mod@primitivesurvival_3.4.7.zip
mod@CarryOn-1.19_v1.7.4.zip
mod@electricity_0.0.11.zip
mod@medievalexpansion-3.13.1.zip
mod@rivers_2.5.0.zip						mod@rivers_2.7.0.zip
mod@StoneRailings.1.2.1.zip

# Maybe shared
mod@joyofsailing-1.3.0.zip					mod@josailing-1.2.8.zip
```

My current guess is that it is related to `electricity`, too, as the chunk we had missing did contain motors from that mod. Though I couldn't find anything that should mess that deeply with the mechanical power system of VS that deeply (from a quick glance at the source code of that mod.

It could also be related to people joining in this chunk. I believe we have electric power generation (generators) in a different chunk nearby.

I hope this helps.

**Comment by Craluminum2413** [comment]
Closing due to inactivity

**Comment by sekelsta** [comment]
Fixed by Radfast, apparently accidentally, last week. Fix released in 1.22.0-rc.9
